blog/_posts/2019-01-18-fluxdiffeq.md

```diff
@@ -247,18 +209,26 @@ as `θ` in associated literature). The utility of this will be seen later.
 
 ## Let's Put an ODE Into a Neural Net Framework!
 
-To understand embedding an ODE into a neural network, let's recall what a layer
-of a neural network actually is. A layer is a transformation 𝐑ⁿ→𝐑ᵐ which
-takes in a vector of size `n` and spits out a new vector of size `m`. That's it!
-So a differential equation layer is simply a layer that takes in a parameter
-vector `p` (of size `n`) and spits out a new vector (of size `m`). You can do this
-manually using the
-[`remake` function in DifferentialEquations.jl](http://docs.juliadiffeq.org/latest/basics/problem.html#Modification-of-problem-types-1),
-or you can use the Flux.jl layers defined in DiffEqFlux.jl. Let's give those a spin.
-
-The most basic differential equation layer is `diffeq_rd`. The function is of
-the form `diffeq_rd(p,reduction,prob,args...;kwargs...)`. The way to understand this
-function is that it does the following. It takes in the parameter vector `p`,
+To understand embedding an ODE into a neural network, let's look at what a
+neural network layer actually is. A layer is really just a *differentiable
+function* which takes in a vector of size `n` and spits out a new vector of size
+`m`. That's it! Layers have traditionally been simple functions like matrix
+multiply, but in the spirit of [differentiable
+programming](https://julialang.org/blog/2018/12/ml-language-compiler) people are
+increasingly experimenting with much more complex functions, such as ray tracers and
+physics engines.
+
+Turns out that differential equations solvers fit this framework, too: A solve
+take in some vector `p` (which might include parameters like the initial
+starting point), and outputs some new vector, the solution. Moreover it's
+differentiable, which means we can put it straight into a larger differentiable

```

Comment:
not necessarily. The examples here, yes.

Reply:
Again, we can keep this as a simplification

Reply:
A later part of the post says:
> The core technical challenge: backpropagation through differential equation solvers. Let's end by explaining the technical issue that needed a solution to make this
all possible. The core to any neural network framework is the ability to
backpropagate derivatives in order to calculate the gradient of the loss function
with respect to the network's parameters. Thus if we stick an ODE solver as a
layer in a neural network, we need to backpropagate through it.

that seems inconsistent with "differentiability isn't necessary". What's your alternative way of combining neural nets and diffeqs?

Reply:
SDE solutions aren't differentiable because they are only Holder continuous with coefficient < 1/2 almost surely, but they train just fine. You don't need differentiability to have an idea of Taylor expansions and weak derivatives. This is a discussion for another time though.

Reply:
Also, a lot of the problems we are solving, even here, are only differentiable a.e.

Reply:
I mean, OK, but that's clearly not how I'm using the word here. We can perhaps clarify in a footnote that we mean differentiability in the AD/ML sense, e.g. "weak piecewise differentiable" or whatever you think is best.